# HG changeset patch
# User Maxim Dounin <mdounin@mdounin.ru>
# Date 1315324342 -14400
# Node ID e165452c1be8865046f3b498c49285e459b8a0c3
# Parent  72e3bfd09f58be13e6d6ca736730eaf4ab994ca9
Incorrect special case for "return 204" removed.
The special case in question leads to replies without body in
configuration like
    location / { error_page 404 /zero; return 404; }
    location /zero { return 204; }
while replies with empty body are expected per protocol specs.
Correct one will look like
    if (status == NGX_HTTP_NO_CONTENT) {
        rc = ngx_http_send_header(r);
        if (rc == NGX_ERROR || r->header_only) {
            return rc;
        }
        return ngx_http_send_special(r, NGX_HTTP_LAST);
    }
though it looks like it's better to drop this special case at all.
diff --git a/src/http/ngx_http_core_module.c b/src/http/ngx_http_core_module.c
--- a/src/http/ngx_http_core_module.c
+++ b/src/http/ngx_http_core_module.c
@@ -1790,11 +1790,6 @@ ngx_http_send_response(ngx_http_request_
 
     r->headers_out.status = status;
 
-    if (status == NGX_HTTP_NO_CONTENT) {
-        r->header_only = 1;
-        return ngx_http_send_header(r);
-    }
-
     if (ngx_http_complex_value(r, cv, &val) != NGX_OK) {
         return NGX_HTTP_INTERNAL_SERVER_ERROR;
     }
